source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -477,6 +493,19 @@ void ContextImpl::logHandshake(SSL* ssl) const {
   }
 }
 
+std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMethodProviders() {
+  std::vector<Envoy::Ssl::PrivateKeyMethodProviderSharedPtr> providers;
+
+  for (uint i = 0; i < tls_contexts_.size(); i++) {
+    Envoy::Ssl::PrivateKeyMethodProviderSharedPtr provider =
+        tls_contexts_[i].getPrivateKeyMethodProvider();
```

Comment:
How does multiple providers per context work?

Reply:
I got what you mean now -- I was first thinking you meant multiple certificates support. The code to tie the providers to the contexts is a remnant from the original design when the providers were context-specific. I need to figure out the best way to sort this out.

Reply:
It turns out that the code was correct after all. The `tls_contexts_` vector refers to `SSL_CTX` objects associated with a given context, and not to Envoy-level contexts as I first thought.

The function here gathers all private key method providers associated with a given TLS context into a vector which is then passed to `ssl_socket.cc`. The code in the ssl socket associates all the providers in the list with the ssl socket, so that if the `SSL_CTX` is changed during handshake in the multi-cert case, the callback to ssl socket will still work.

I added some tests (and a very limited ECDSA private key method provider) to validate this case. If you have any particular configuration in mind which should be tested, please tell, and I'll try it out and add a test for it.